site/content/blog/2019-10-25-instrumenting-go-code-via-ast.md

```diff
@@ -0,0 +1,315 @@
+---
+title: "Instrumenting Go via AST"
+slug: instrumenting-go-code-via-ast
+date: 2019-10-25T00:00:00-04:00
+author: Eli Yukelzon
+github: reflog
+community: eli.yukelzon
+---
+
+We've been working on integrating call tracing in the server to provide exact measurements of all API and DB calls.
+We've picked [OpenTracing](https://github.com/opentracing/opentracing-go) - a lovely open source project that allows you to setup trace reporting and enables you to support [Distributed tracing](https://opentracing.io/docs/overview/what-is-tracing/)
+
+Instrumenting your API handler in Go is very straightforward - setup connection to a collection server supporting OpenTracing (we've decided to use [Jaeger](https://www.jaegertracing.io/)) spec and wrap you code in spans.
+
+To simplify this, we've added a simple tracing module:
+```go
+package tracing
+
+import (
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/uber/jaeger-lib/metrics"
+
+	"github.com/uber/jaeger-client-go"
+	jaegercfg "github.com/uber/jaeger-client-go/config"
+	jaegerlog "github.com/uber/jaeger-client-go/log"
+	"context"
+)
+
+var initialized = false
+
+func Initialize() error{
+	cfg := jaegercfg.Configuration{
+		Sampler: &jaegercfg.SamplerConfig{
+			Type:  jaeger.SamplerTypeConst,
+			Param: 1,
+		},
+		Reporter: &jaegercfg.ReporterConfig{
+			LogSpans: true,
+		},
+	}
+
+	_, err := cfg.InitGlobalTracer(
+		"mattermost",
+		jaegercfg.Logger(jaegerlog.StdLogger),
+		jaegercfg.Metrics( metrics.NullFactory),
+	)
+	if err != nil {
+		return err
+	}
+
+	initialized = true
+
+	return nil
+}
+
+func StartRootSpanByContext(ctx context.Context, operationName string)(opentracing.Span, context.Context) {
+	return opentracing.StartSpanFromContext(ctx, operationName)
+}
+
+func StartSpanWithParentByContext(ctx context.Context, operationName string) (opentracing.Span, context.Context) {
+	parentSpan := opentracing.SpanFromContext(ctx)
+
+	if parentSpan == nil {
+		return StartRootSpanByContext(ctx, operationName)
+	}
+
+	return opentracing.StartSpanFromContext(ctx, operationName, opentracing.ChildOf(parentSpan.Context()))
+}
+```
+
+
+Now, wrapping an API call in a span is very easy:
+
+```go
+
+func apiCall(c *Context, w http.ResponseWriter, r *http.Request) {
+	span, ctx := tracing.StartSpanWithParentByContext(c.App.Context, "api4:apiCall")
+	c.App.Context = ctx
+    defer span.Finish()
+
+    // perform actual request handling
+}
+```
+
+Now, each time the `apiCall` handler is invoked, it will be measured and traced on Jaeger.
+
+The problem is - we have a rather large API surface (around 300+ handlers) and it would be a really tough task to instrument all the handlers by hand. That's where our story begins.
```

Comment:
```suggestion
The problem is: we have a rather large API surface (around 300+ handlers) and it would be a really tough task to instrument all the handlers by hand. That's where our story begins.
```